ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/transaction/UtxoTransactionBuilder.java

```diff
@@ -180,21 +180,29 @@ public interface UtxoTransactionBuilder {
     Map<String, List<ContractState>> getEncumbranceGroups();
 
     /**
-     * Gets the notary assigned to the current transaction, or null if the notary has not been set.
+     * Gets the notary service name assigned to the current transaction, or null if the notary has not been set.
      *
-     * @return Returns the notary assigned to the current transaction, or null if the notary has not been set.
+     * @return Returns the service name of the notary assigned to the current transaction, or null if the notary has not been set.
      */
     @Nullable
-    Party getNotary();
+    MemberX500Name getNotaryName();
 
     /**
-     * Sets the specified {@link Party} as a notary to the current {@link UtxoTransactionBuilder}.
+     * Gets the notary service key assigned to the current transaction, or null if the notary has not been set.
      *
-     * @param notary The {@link Party} to set as a notary to the current {@link UtxoTransactionBuilder}.
+     * @return Returns the service key of the notary assigned to the current transaction, or null if the notary has not been set.
+     */
+    @Nullable
+    PublicKey getNotaryKey();
+
+    /**
+     * Sets the specified {@link MemberX500Name} as a notary to the current {@link UtxoTransactionBuilder}.
+     *
+     * @param notary The {@link MemberX500Name} to set as a notary to the current {@link UtxoTransactionBuilder}.
      * @return Returns a new {@link UtxoTransactionBuilder} with the new notary.
      */
     @NotNull
-    UtxoTransactionBuilder setNotary(@NotNull Party notary);
+    UtxoTransactionBuilder setNotary(@NotNull MemberX500Name notary);
```

Comment:
Not sure this is right decision. Should we have the key as primary data and the name is simply derived from Group Parameters?

Reply:
Is there a reason that the x500 name doesn't work? This API will be used by CorDapps. In the case when they need to select a notary service (because there are multiple on the network), wouldn't their business logic be based on the x500 name? Keys can change so they can't really lock those into their lookup, they'd have to use the group parameters to get the key for the name they want anyway.

Reply:
I agree with Ramzi here - flows should select the notary by X500 name, we look up the current key from the Group Parameters (that might still need doing). After that, we can store the name and the key in the transaction and sign over it, so we can check that the expected key was used, and we can use historic group parmeters to check that the key was valid at the time of notarisation.

Reply:
Actually, the notary key gets looked up from the group params via the notary look-up, so that bit is fine.

Reply:
You convinced me